### PR TITLE
`Pagination` main functionality (HDS-969)

### PR DESCRIPTION
**Pull request by KristinLBradley**
### :pushpin: Summary
Implements functionality for **compact** and basic **numbered** pagination.

**Preview:** https://hds-components-git-hds-969-pagination-functionality-hashicorp.vercel.app/components/pagination-bar#showcase

### :hammer_and_wrench: Detailed description
* Functionality for the "compact" version of Pagination including:
  * Prev & Next buttons for navigating through pages
  * `TotalCount` (displays current items range and total item count)
  * `PageSize` (select number of items per page to display)
* Basic functionality for "numbered" version of Pagination including:
  * list of page links
  * setting selected page
  * Does NOT include truncation for numbered version, will be done in separate branch
* This branch will be merged into the https://github.com/hashicorp/design-system/pull/661 once approved. (It will not be merged directly into the main branch.)
* CSS and documentation will be cleaned up and refined in separate branches

### :link: External links
- https://hashicorp.atlassian.net/browse/HDS-969

***

### 👀 How to review

Would like feedback on my general approach in terms of how the sub-components work together with the PaginationBar wrapper component as well as the functionality implemented so far.

👉 Review commit-by-commit
👉 Review by files changed

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

File: packages/components/addon/components/hds/pagination-bar/index.js
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/index.js:13]: probably we may want to expose the possibility for the consumers to start from a specific page? (for example they store the current page in an URL, and the user reloads the page when `currentPage` is `3`)

File: packages/components/addon/components/hds/pagination-bar/index.js
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/index.js:15]: suggestion (based on experience): add as many comments as you can, when there is this kind of complex math/logic: in six months from now, I assure you, you will not remember why you did something in a certain way, what does it mean, what happens if you need to change something, etc.

make your future self happy 🙂  (and the people that will work on this code after you, too 😉)

File: packages/components/addon/components/hds/pagination-bar/index.js
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/index.js:19]: see suggestion above about adding comments

File: packages/components/addon/components/hds/pagination-bar/pagination.hbs
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/pagination.hbs:6]: Probably I would abstract this straight away in its own component, so I can:
- re-use it/share it between `prev` and `next`
- test it in isolation (the API, the UI, the interaction)
- add it in the showcase, with its different variants and states

File: packages/components/addon/components/hds/pagination-bar/pagination.js
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/pagination.js:41]: in case you want to avoid the duplication of code between `prevPage` and `nextPage` you could consider to use a generic method `changePage(direction)`, where `direction` (can be `prev` or `next`) controls only the change in the value (+1 or -1)

File: packages/components/app/styles/components/pagination-bar.scss
Comment (didoo) [inline packages/components/app/styles/components/pagination-bar.scss:56]: I haven't looked at the CSS yet. Let me know if want me to start to have a look.

**Review by didoo — COMMENTED** [review]
@KristinLBradley added a few comments, but I want to chat more in details so I've set up a meeting for Monday, so we have more time.

File: packages/components/app/styles/components/pagination-bar.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/pagination-bar.scss:56]: Thanks, I'm going through your other comments but the CSS is not yet ready (I need to work on layout differences for different combinations of sub-components after this).

File: packages/components/addon/components/hds/pagination-bar/pagination.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/pagination.hbs]: Refactors are still a WIP but I separated out the buttons now and pushed that up. I will work on updating the showcase, etc. later on.

File: packages/components/tests/dummy/app/templates/components/pagination-bar.hbs
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/pagination-bar.hbs:158]: Commented this out until it's working (was getting an error)

File: packages/components/addon/components/hds/pagination-bar/index.hbs
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/index.hbs:6]: you don't need a specific `hasTotalItems` argument, you can rely on the existence (not `undefined` of `@totalItems` to know if it has total items or not).

File: packages/components/addon/components/hds/pagination-bar/index.hbs
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/index.hbs:28]: super nitpick: in general we tend to have arguments/props first, and then event handlers
```suggestion
      PageSize=(component "hds/pagination-bar/page-size" itemsPerPage=this.itemsPerPage onChange=this.onPageSizeChange)
```

for example: https://github.com/hashicorp/design-system/blob/main/packages/components/addon/components/hds/tabs/index.hbs#L9-L14

File: packages/components/addon/components/hds/pagination-bar/page-size.hbs
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/page-size.hbs:2]: since the typographic styles should only apply to the label, better to move them where their actual application context is:
```suggestion
<div class="hds-pagination-page-size" ...attributes>
  <label for={{this.pageSizeId}} class="hds-pagination-page-size hds-typography-body-200 hds-font-weight-medium">Items per page</label>
```

File: packages/components/app/styles/components/pagination-bar.scss
Comment (didoo) [inline packages/components/app/styles/components/pagination-bar.scss:142]: if you move the typographic classes to the `label` then you have to remove these two rules

File: packages/components/addon/components/hds/pagination-bar/page-size.hbs
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/page-size.hbs:3]: I see two "on change" handlers applied to the same element, probably we can unify them in a single call on the backing class. I can show you how if you want.

File: packages/components/addon/components/hds/pagination-bar/total-count.js
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/total-count.js:17]: as mentioned above, you may consider doing this instead (unless I am not misunderstanding something):
```suggestion
    return this.args.totalItems !== undefined;
```
and get rid of the `@hasTotalItems` parameter

File: packages/components/addon/components/hds/pagination-bar/page-size.js
Comment (didoo) [inline packages/components/addon/components/hds/pagination-bar/page-size.js:11]: see? we were not "saving" that much time/code 🙂 (on the contrary we were introducing overrides/special cases in our design system)

**Review by didoo — COMMENTED** [review]
@KristinLBradley I've done a first pass at part of the code, and left a few comments. Overall, what I'm seeing is great! Nothing major to point out, from the "code architecture" perspective you're spot on (happy to see how quickly you got to use the contextual components patter, and how things are passed up-down between parent-child elements). Can't wait to see the whole Pagination component code done 🚀

Notice: I have **not**  looked into the `PaginationBar::Pagination` and its sub-components (`Pagination::NavButton`, `Pagination::PageNumber`) because I am not sure what is ready and what you're still working on; same thing for the CSS, which I think is still a WIP. Let me know if there is something of what I've left out that you want me to look into, OK?

File: packages/components/addon/components/hds/pagination-bar/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/index.hbs:6]: My understanding of whether to show the total number of items (hasTotalItems) was that it was just an option for the end user to customize the visual UI. 

For example, displaying "1–10" vs. "1–10 of 100". By default it seems it should be true.

Maybe the name is confusing though and we could come up with a clearer name?

**Comment by KristinLBradley** [comment]
> @KristinLBradley I've done a first pass at part of the code, and left a few comments. Overall, what I'm seeing is great! Nothing major to point out, from the "code architecture" perspective you're spot on (happy to see how quickly you got to use the contextual components patter, and how things are passed up-down between parent-child elements). Can't wait to see the whole Pagination component code done 🚀
> 
> Notice: I have **not** looked into the `PaginationBar::Pagination` and its sub-components (`Pagination::NavButton`, `Pagination::PageNumber`) because I am not sure what is ready and what you're still working on; same thing for the CSS, which I think is still a WIP. Let me know if there is something of what I've left out that you want me to look into, OK?

Thanks! If you have feedback on other things you think would make sense as child components or things like that I would like to know. I still need to clean up the CSS a bit so I will do that later on. I will also do other clean up by adding more comments in the JS but if you have general comments related to that, that's fine.

After this PR is merged I plan to work on more of that minor clean up in addition to the truncation.

File: packages/components/addon/components/hds/pagination-bar/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/index.hbs:28]: Thanks, I still need to clean up and reorganize some of the code like this but this is helpful.

File: packages/components/addon/components/hds/pagination-bar/page-size.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/page-size.hbs]: Thanks, I'll look at these. I haven't started implementing responsive layouts yet.

File: packages/components/addon/components/hds/pagination-bar/total-count.js
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/total-count.js:17]: See my other comments above. "hasTotalItems" is just a UI option to show or hide the total # of items in the UI. (We should consider clearer naming though...)

File: packages/components/addon/components/hds/pagination-bar/total-count.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/total-count.hbs:2]: I can look at this, thanks

File: packages/components/addon/components/hds/pagination-bar/page-size.js
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/page-size.js:11]: Sure I see.

File: packages/components/addon/components/hds/pagination-bar/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/index.hbs:6]: I agree it's confusing. I will work more on the documentation after this.

File: packages/components/addon/components/hds/pagination-bar/total-count.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/total-count.hbs:2]: I think I'll leave it for this PR. I'm worried it will make the template more confusing.

File: packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs
Comment (KristinLBradley) [inline packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs:2]: @didoo let me know if you think this (and the `pageNumber` child components) should just use html `button` tags instead. (I used `Hds::Interactive` because of previous feedback)

File: packages/components/tests/dummy/app/templates/components/pagination-bar.hbs
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/pagination-bar.hbs:254]: @didoo I commented these examples out because `@totalItems` is required. Let me know if it's ok to delete these or if there are other refactors you think should be made.

File: packages/components/addon/components/hds/pagination-bar/index.hbs
Comment (didoo) [reply packages/components/addon/components/hds/pagination-bar/index.hbs:6]: Resolving the conversation for now, we can review this later.

File: packages/components/tests/dummy/app/templates/components/pagination-bar.hbs
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/pagination-bar.hbs:280]: something in these examples is not working, because I don't see the "disabled" state of the prev/next elements

**Review by didoo — COMMENTED** [review]
@KristinLBradley I've added a lot of comments, but nothing major: as mentioned before, the overall approach to the component architecture/organization is OK, we have to work on a few details (I've set up another meeting for today, maybe we can do some pair programming and iron out some of them)

File: packages/components/addon/components/hds/pagination-bar/total-count.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/total-count.hbs:3]: Thanks, I will change it

File: packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs:11]: Makes sense, thanks. I will change it.

File: packages/components/tests/dummy/app/templates/components/pagination-bar.hbs
Comment (KristinLBradley) [reply packages/components/tests/dummy/app/templates/components/pagination-bar.hbs:190]: Thanks, I just tried merging in the latest main but still saw the error. I will try again later on.

File: packages/components/tests/dummy/app/templates/components/pagination-bar.hbs
Comment (KristinLBradley) [reply packages/components/tests/dummy/app/templates/components/pagination-bar.hbs:244]: I'm seeing the disabled states as expected. Are you positive you have been looking at the latest updates? (I pushed all the updates I made from your previous round of feedback Monday.)

Screen Shot 2022-12-06 at 10 04 14 AM

File: packages/components/addon/components/hds/pagination-bar/index.js
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/index.js:23]: Ok, I can take a look later.

File: packages/components/addon/components/hds/pagination-bar/index.js
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/index.js:52]: Well when you're changing the page size you don't know exactly where you are after that so going back to the first page seems to make sense.

File: packages/components/addon/components/hds/pagination-bar/index.js
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/index.js]: Updated the name and made it tracked

File: packages/components/addon/components/hds/pagination-bar/index.js
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/index.js]: I created a generic function for it

File: packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs]: Sorry missed that one. I removed.

File: packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs:9]: I decided instead to just clarify the documentation since currently the NavButton component is only used internally and I didn't know if it made sense to add another parameter to the Pagination wrapper component. We can reexamine later on though.

File: packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs:4]: There was an issue with this I fixed. I can look into possibly using a getter later

File: packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/pagination/nav-button.hbs:11]: I'll consider this when I focus on styling again later on.

File: packages/components/addon/components/hds/pagination-bar/pagination/page-number.js
Comment (KristinLBradley) [reply packages/components/addon/components/hds/pagination-bar/pagination/page-number.js]: Ok, I moved the logic

**Review by didoo — APPROVED** [review]
All good to go and merge (once you've checked again things, after our pairing session :) )

**Pull request merged**
